### PR TITLE
Binary trie

### PR DESCRIPTION
**Pull request by ajlopezrsk**
The support for other arities (number of subnodes per node) (4, 16,..) was removed. Tests using arity 16 were removed too.

Tested againts testnet and mainnet (some thoushands of blocks) without breaking consensus.

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 20 issues
* ![MAJOR][MAJOR] 14 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 5 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [TrieConstants.java#L24](https://github.com/rsksmart/rskj/blob/ddbe8ec2b2a57e68ae36640a5d550bbbbaa48d18/rskj-core/src/main/java/co/rsk/trie/TrieConstants.java#L24): Add a private constructor to hide the implicit public one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [TrieImpl.java#L186](https://github.com/rsksmart/rskj/blob/ddbe8ec2b2a57e68ae36640a5d550bbbbaa48d18/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L186): The Cyclomatic Complexity of this method "fromMessage" is 13 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [TrieImpl.java#L252](https://github.com/rsksmart/rskj/blob/ddbe8ec2b2a57e68ae36640a5d550bbbbaa48d18/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L252): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [TrieImpl.java#L444](https://github.com/rsksmart/rskj/blob/ddbe8ec2b2a57e68ae36640a5d550bbbbaa48d18/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L444): The Cyclomatic Complexity of this method "toMessage" is 12 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [TrieImpl.java#L572](https://github.com/rsksmart/rskj/blob/ddbe8ec2b2a57e68ae36640a5d550bbbbaa48d18/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L572): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L576](https://github.com/rsksmart/rskj/blob/ddbe8ec2b2a57e68ae36640a5d550bbbbaa48d18/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L576): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L588](https://github.com/rsksmart/rskj/blob/ddbe8ec2b2a57e68ae36640a5d550bbbbaa48d18/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L588): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L675](https://github.com/rsksmart/rskj/blob/ddbe8ec2b2a57e68ae36640a5d550bbbbaa48d18/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L675): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L681](https://github.com/rsksmart/rskj/blob/ddbe8ec2b2a57e68ae36640a5d550bbbbaa48d18/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L681): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L901](https://github.com/rsksmart/rskj/blob/ddbe8ec2b2a57e68ae36640a5d550bbbbaa48d18/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L901): The Cyclomatic Complexity of this method "put" is 11 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/co/rsk/trie/TrieImpl.java [hunk 1]
```diff
@@ -211,14 +193,19 @@ private static TrieImpl fromMessage(byte[] message, int position, int msglength,
 
         try {
             int arity = istream.readByte();
+
+            if (arity != 2) {
```
Comment (tinchou) [inline rskj-core/src/main/java/co/rsk/trie/TrieImpl.java]: use the `ARITY` constant

File: rskj-core/src/main/java/co/rsk/trie/TrieConstants.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.trie;
+
+/**
+ * Created by ajlopez on 27/01/2018.
+ */
+public class TrieConstants {
```
Comment (tinchou) [inline rskj-core/src/main/java/co/rsk/trie/TrieConstants.java]: This is only used in `TrieImpl`. Please make it a private constant of that class.

**Comment by rskops** [comment]
SonarQube analysis reported 19 issues
* ![MAJOR][MAJOR] 13 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 5 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [TrieImpl.java#L186](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L186): The Cyclomatic Complexity of this method "fromMessage" is 13 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [TrieImpl.java#L252](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L252): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [TrieImpl.java#L444](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L444): The Cyclomatic Complexity of this method "toMessage" is 12 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [TrieImpl.java#L572](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L572): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L576](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L576): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L588](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L588): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L675](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L675): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L681](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L681): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L901](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L901): The Cyclomatic Complexity of this method "put" is 11 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [TrieImpl.java#L928](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L928): Possible null pointer dereference of firstChild in co.rsk.trie.TrieImpl.put(byte[], int, int, byte[]) [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 19 issues
* ![MAJOR][MAJOR] 13 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 5 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [TrieImpl.java#L188](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L188): The Cyclomatic Complexity of this method "fromMessage" is 13 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [TrieImpl.java#L254](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L254): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [TrieImpl.java#L446](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L446): The Cyclomatic Complexity of this method "toMessage" is 12 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [TrieImpl.java#L574](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L574): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L578](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L578): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L590](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L590): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L677](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L677): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L683](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L683): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [TrieImpl.java#L903](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L903): The Cyclomatic Complexity of this method "put" is 11 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [TrieImpl.java#L930](https://github.com/rsksmart/rskj/blob/6258071bc5da589980f701a5df30194645363803/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L930): Possible null pointer dereference of firstChild in co.rsk.trie.TrieImpl.put(byte[], int, int, byte[]) [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**